src/rdkafka_broker.c

```diff
@@ -3631,13 +3633,74 @@ static void rd_kafka_toppar_fetch_backoff (rd_kafka_broker_t *rkb,
                 return;
 
         rktp->rktp_ts_fetch_backoff = rd_clock() + (backoff_ms * 1000);
+
         rd_rkb_dbg(rkb, FETCH, "BACKOFF",
-                   "%s [%"PRId32"]: Fetch backoff for %dms: %s",
+                   "%s [%"PRId32"]: Fetch backoff for %dms%s %s",
                    rktp->rktp_rkt->rkt_topic->str, rktp->rktp_partition,
-                   backoff_ms, rd_kafka_err2str(err));
+                   backoff_ms,
+                   err ? ":" : "", 
+                   err ? rd_kafka_err2str(err) : "");
 }
 
 
+/**
+ * @brief Handle preferred replica in fetch response.
+ * 
+ * @locks rd_kafka_toppar_lock(rktp) must be held.
+ *        rd_kafka_rdlock(rk) must NOT be held.
+ *
+ * @locality broker thread
+ */
+void
+rd_kafka_fetch_preferred_replica_handle (rd_kafka_toppar_t *rktp,
+                                         rd_kafka_buf_t *rkbuf,
+                                         rd_kafka_broker_t *rkb,
+                                         int32_t preferred_id) {
+        rd_kafka_broker_t *preferred_rkb;
+        rd_kafka_t *rk = rktp->rktp_rkt->rkt_rk;
+        rd_ts_t one_minute = 60*1000*1000;
+        rd_ts_t five_seconds = 5*1000*1000;
+        
+        if (rd_interval_immediate(&rktp->rktp_new_lease_intvl,
+                                  one_minute, 0) < 0)
+                /* In lieu of KIP-320, the toppar is delegated back to
+                 * the leader in the event of an offset out-of-range 
+                 * error (KIP-392 error case #4) because this scenario
+                 * implies the preferred replica is out-of-sync.
+                 * 
+                 * If program execution reaches here, the leader has
+                 * relatively quickly instructed the client back to 
+                 * a preferred replica, quite possibly the same one
+                 * as before (possibly resulting from stale metadata),
+                 * so we back off the toppar to slow down potential
+                 * back-and-forth.
+                 */
+                rd_kafka_toppar_fetch_backoff(rkb,
+                        rktp, RD_KAFKA_RESP_ERR_NO_ERROR);
+        
+        rd_kafka_rdlock(rk);
```

Comment:
wonderful plan, this is was indeed an issue as i've no doubt you were aware.

Reply:
i've changed this to only lock the rktp over the `rd_kafka_toppar_broker_update` call.

please confirm locks aren't necessary over the `rd_kafka_toppar_fetch_backoff` calls (since they're all on the broker thread)